### ISSUE TITLE
Is HW transcoding really working?

### ISSUE THREAD

**Issue by ahiguero** [linuxserver/docker-plex#222]
[linuxserverurl]: https://linuxserver.io
[![linuxserver.io](https://raw.githubusercontent.com/linuxserver/docker-templates/master/linuxserver.io/img/linuxserver_medium.png)][linuxserverurl]

If you are new to Docker or this application our issue tracker is **ONLY** used for reporting bugs or requesting features. Please use [our discord server](https://discord.gg/YWrKVTn) for general support.

running the docker container on a QNAP TS-453A with the following docker setup

docker create \
  --name=plex \
  --net=host \
  --device=/dev/dri/card0:/dev/dri/card0 \
  --device=/dev/dri/renderD128:/dev/dri/renderD128 \
  -e PUID=1002 \
  -e PGID=100 \
  -e TZ=America/Chicago \
  -e VERSION=latest \
  -v /share/CACHEDEV1_DATA/cs_config/plex_config:/config \
  -v /mnt/rf/nd/0/02ca9787e-904f-42c1-99a6-c38b52d1128b/tv_series:/tv \
  -v /mnt/rf/nd/0/02ca9787e-904f-42c1-99a6-c38b52d1128b/movies:/movies \
  -v /mnt/rf/nd/0/02ca9787e-904f-42c1-99a6-c38b52d1128b/hires_audio:/music \
  -v /mnt/rf/nd/0/02ca9787e-904f-42c1-99a6-c38b52d1128b/drone:/drone \
  -v /mnt/rf/nd/0/02ca9787e-904f-42c1-99a6-c38b52d1128b/preroll:/preroll \
  -v /dev/shm:/transcode \
  --restart unless-stopped \
  linuxserver/plex
------------------------------

## Expected Behavior
the setup should use the HW transcoding as explained in the instructions provided int he read me

## Current Behavior
the only way to get hw transcoding to work is by issuing the following commands in the NAS terminal

chmod 777 /dev/dri /dev/dri/card0 /dev/dri/renderD128

## Steps to Reproduce
1. install docker based not eh instructions.
2. try to not use direct play or the new render engine not eh plex client / or set home streaming quality to lets say 720p
3.
4.

QNAP NAS TS-453A

**Comment by aptalca** [comment]
map `/dev/dri` as the instructions suggest

**Comment by ahiguero** [comment]
The mapping as suggested was done --device=/dev/dri:/dev/dri and still was not working until the chmod 777 /dev/dri /dev/dri/card0 /dev/dri/renderD128 command was issue

**Comment by aptalca** [comment]
What were the original ownership and perms prior to chmod?

**Comment by ahiguero** [comment]
Here is the out of the ls command on the NAS
[~] # ls -al /dev/dri
total 0
drwxr-xr-x  2 admin administrators       80 2020-02-13 09:09 ./
drwxr-xr-x 15 admin administrators    20740 2020-02-13 15:17 ../
crw-------  1 admin administrators 226,   0 2020-02-13 09:09 card0
crw-------  1 admin administrators 226, 128 2020-02-13 09:09 renderD128

**Comment by aptalca** [comment]
sigh, it seems qnap doesn't really follow any udev rules.

typically, those render devices are owned by a video group, and they have group r/w permissions so that any process that belongs to that video group can access and use them. 

We have logic inside our container that adds the plex user to such video group. Unfortunately here, not only there is no video group (but administrators), and the device don't have any group r/w permissions.

The only way to get them to work is by chmod'ing, on the host, which is what you did. The ideal solution would be for qnap to put them under a video group and give them group r/w perms.

There isn't much we can do, it's a qnap restriction.

**Comment by aptalca** [comment]
out of curiosity, can you check the admin user id? `id admin`

**Comment by ahiguero** [comment]
[/] # id admin
uid=0(admin) gid=0(administrators) groups=0(administrators),100(everyone)

**Comment by billerr** [comment]
Sorry for resurrecting an old thread, but I fixed the above issue in a clean (I think) way, and wanted to share.

In my `/config/custom-cont-init.d` directory, I added the following `10-permissions` script: 
```bash
#!/usr/bin/with-contenv bash

# fix device group permissions
echo '[10-permissions] *** Fixing /dev/dri permissions ***'
chmod g+rw /dev/dri/*
```

Some notes:
- I added the file from the Qnap shell as user `admin`, to match the permissions of the directory (which is created by Container Station).
- Since the script runs inside the container, the devices keep their default restrictive permissions in the Qnap environment, and our changes only show within the container. Win-win I guess?

@aptalca could this also be added to the image (e.g. in the `50-gid-video` script) to make sure the devices always have proper permissions in the container?

**Comment by aptalca** [comment]
Iirc there was a reason why we didn't do that but I can't remember what it was. If the device is missing group rw perms, the log should print a notice and a command to fix it on the host

**Comment by billerr** [comment]
Oh ok - so far it works alright, but I'll watch out for any side-effects and update the thread.

**Comment by jamesking420** [comment]
This issue has also been impacting me, and I have found that to fix it I have done 
```
chmod /666 /dev/dri/*
```
in an autorun.sh script on boot.  

I see the solution provided by @billerr [here](https://github.com/linuxserver/docker-plex/issues/222#issuecomment-1003420410), which I would love to follow, since it seems like a cleaner method but I am not confident enough to do it.

Do I just need to create the file called 10-permissions (or 10-permission.sh?) in order to do this?  With the exact content provided by @billerr ?

**Comment by aptalca** [comment]
Did you check the log? It gives you the exact command to run on the host, which is the cleanest method. Run that command and all your containers that need access to that device will work.

If your host is somehow reverting those permissions during updates, put that in the auto run or whatever script that runs on restart of the host.

**Comment by jamesking420** [comment]
It is just booting up again now, and I'll take a look.  You mean the log for the container?

**Comment by jamesking420** [comment]
PS - Possibly unrelated but I did previously notice that my log shows this: ```Critical: libusb_init failed ``` Don't know if it's important!

**Comment by jamesking420** [comment]
Unfortunately, no such information on what to do in the docker logs @aptalca - at least, not that I understand.
https://www.toptal.com/developers/hastebin/ajijosicag.yaml

**Comment by aptalca** [comment]
did you map `/dev/dri` as a device? If so, then the permissions are correct.

https://github.com/linuxserver/docker-plex/blob/master/root/etc/cont-init.d/50-gid-video#L17

**Comment by jamesking420** [comment]
I did yes, in my docker-compose file.  However, HW transcoding does not work without explicitly updating the permissions.

I just got @billerr script working and HW transcoding is now functioning.  It might be that this is to do with the admin user being user group 0? (I saw something about this elsewhere, though Ican't find it now.

I either need @billerr's script, or to add similar to autorun.sh.

The more I have investigated this, the more it seems to be a fairly crappy QNAP permissions issue.  I am happy to debug it, or, given I have it working, I will not bother you further if you are not interested. :)
